src/local_scheduler/local_scheduler_algorithm.cc

```diff
@@ -112,16 +110,29 @@ struct SchedulingAlgorithmState {
   object_entry *remote_objects;
 };
 
+TaskQueueEntry TaskQueueEntry_init(TaskSpec *spec, int64_t task_spec_size) {
+  TaskQueueEntry elt;
+  elt.spec = (TaskSpec *) malloc(task_spec_size);
+  memcpy(elt.spec, spec, task_spec_size);
+  elt.task_spec_size = task_spec_size;
+  return elt;
+}
+
+void TaskQueueEntry_free(TaskQueueEntry *entry) {
+  TaskSpec_free(entry->spec);
+}
+
 SchedulingAlgorithmState *SchedulingAlgorithmState_init(void) {
   SchedulingAlgorithmState *algorithm_state =
       (SchedulingAlgorithmState *) malloc(sizeof(SchedulingAlgorithmState));
+  memset(algorithm_state, 0, sizeof(SchedulingAlgorithmState));
   /* Initialize an empty hash map for the cache of local available objects. */
   algorithm_state->local_objects = NULL;
   /* Initialize the hash table of objects being fetched. */
   algorithm_state->remote_objects = NULL;
   /* Initialize the local data structures used for queuing tasks and workers. */
-  algorithm_state->waiting_task_queue = NULL;
-  algorithm_state->dispatch_task_queue = NULL;
+  algorithm_state->waiting_task_queue.reset(new std::list<TaskQueueEntry>());
+  algorithm_state->dispatch_task_queue.reset(new std::list<TaskQueueEntry>());
```

Comment:
I haven't seen the whole code, but generally speaking you should (almost) never need `std::unique_ptr<std::vector<T> >`; you should just do `std::vector<T>` and then call `vec1.swap(vec2)` if you need to move them around the vectors for some reason. Note that `swap()` doesn't invalid iterators/pointers/references to elements for `std::vector`, so it's not a problem if that's what you're worried about (this behavior differs from that of `std::basic_string`).